org.osgi.test.common/src/main/java/org/osgi/test/common/service/ServiceConfigurationKey.java

```diff
@@ -1,32 +1,33 @@
 package org.osgi.test.common.service;
 
+import static java.util.Objects.requireNonNull;
+
 import java.util.Arrays;
 import java.util.Objects;
 
-import org.osgi.test.common.annotation.InjectService;
-
-public class ServiceConfigurationKey {
-
-	private final int		cardinality;
-	private final String	filter;
-	private final String[]	filterArguments;
-	private final String	serviceName;
-	private final long		timeout;
-
-	public ServiceConfigurationKey(Class<?> serviceType, InjectService injectService) {
-		this.serviceName = serviceType.getName();
-		this.cardinality = injectService.cardinality();
-		this.filter = injectService.filter();
-		this.filterArguments = injectService.filterArguments();
-		this.timeout = injectService.timeout();
+public class ServiceConfigurationKey<S> {
+
+	final Class<S>	serviceType;
+	final String	filter;
+	final String[]	filterArguments;
+	final int		cardinality;
+	final long		timeout;
+
+	public ServiceConfigurationKey(Class<S> serviceType, String filter, String[] filterArguments, int cardinality,
+		long timeout) {
+		this.serviceType = requireNonNull(serviceType);
+		this.filter = filter;
+		this.filterArguments = filterArguments;
```

Comment:
I think this is done because that  we can use it also as key with the @ServiceSource.

Reply:
I am not sure it is a great idea. We will create a ServiceConfigurationKey each time we go to the Store to get a ServiceConfiguration so we should make the creation of the key as lightweight as possible. Computing the filter when creating the key seems like wasteful work.

Reply:
My thinking was that to use the key, you'll have to call `hashCode()` and `equals()`, which will involve iterating over the filter template and the filter args independently and calling `equals()` and `hashcode()` on each arg. This applies to every `put()` and `get()` operation. If you use a pre-computed filter, then you only have to compute the hashcode and do the equals on the single string.

But without benchmarking we wouldn't know how much of a difference this will make.